service/worker/scanner/workflow/scavenger.go

```diff
@@ -0,0 +1,345 @@
+)
+
+// NewScavenger returns an instance of history scavenger daemon
+// The Scavenger can be started by calling the Run() method on the
+// returned object. Calling the Run() method will result in one
+// complete iteration over all of the history branches in the system. For
+// each branch, the scavenger will attempt
+//  - describe the corresponding workflow execution
+//  - deletion of history itself, if there are no workflow execution
+func NewScavenger(
+	shardDB   p.ShardManager,
+	historyDB p.HistoryManager,
+	resource resource.Resource,
+	rps int,
+	numShards int,
+	hbd ScavengerHeartbeatDetails,
+	metricsClient metrics.Client,
+	logger log.Logger,
+) *Scavenger {
+
+	rateLimiter := rate.NewLimiter(rate.Limit(rps), rps)
+
+	return &Scavenger{
+		shardDB:shardDB,
+		historyDB:historyDB,
+		resource:resource,
+		hbd:     hbd,
+		rps:     rps,
+		numShards:numShards,
+		limiter: rateLimiter,
+		metrics: metricsClient,
+		logger:  logger,
+	}
+}
+
+var persistenceOperationRetryPolicy = common.CreatePersistanceRetryPolicy()
+
+// Run runs the scavenger
+func (s *Scavenger) Run(ctx context.Context) (ScavengerHeartbeatDetails, error) {
+	taskCh := make(chan taskDetail, pageSize)
+	respCh := make(chan taskResult, pageSize)
+	concurrency := s.rps/rpsPerConcurrency + 1
+
+	for i := 0; i < concurrency; i++ {
+		go s.startTaskProcessor(ctx, taskCh, respCh)
+	}
+
+	for {
+		var resp *p.ScanWorkflowsResponse
+		var err error
+		op := func() error {
+			resp, err = s.shardDB.ScanWorkflows(&p.ScanWorkflowsRequest{
+				PageSize:      pageSize,
+				NextPageToken: s.hbd.NextPageToken,
+				ShardID: s.hbd.CurrentShardID,
+			})
+			return err
+		}
+
+		err = backoff.Retry(op, persistenceOperationRetryPolicy, common.IsPersistenceTransientError)
+
+		if err != nil {
+			return s.hbd, err
+		}
+
+		taskCount := 0
+		corruptionCount := 0
+		errorCount := 0
+		for _, wf := range resp.WorkflowExecutions {
+			if wf.IsCurrentWorkflow && p.IsWorkflowRunning(wf.State){
+				taskCount ++
+				taskCh <- taskDetail{
+					shardID:    s.hbd.CurrentShardID,
+					domainID:   wf.DomainID,
+					workflowID: wf.WorkflowID,
+					runID:      wf.RunID,
+					hbd:s.hbd,
+				}
+			}
+		}
+
+		if taskCount > 0 {
+			// wait for task counters indicate this batch is done
+		Loop:
+			for {
+				select {
+				case resp := <-respCh:
+					taskCount --
+					if resp.err != nil {
+						s.metrics.IncCounter(metrics.MutableStateScavengerScope, metrics.MutableStateScavengerErrorCount)
+						errorCount++
+					}
+					if resp.corrupted {
+						s.metrics.IncCounter(metrics.MutableStateScavengerScope, metrics.MutableStateScavengerCorruptionCount)
+						corruptionCount++
+					}
+					if taskCount == 0 {
+						break Loop
+					}
+				case <-ctx.Done():
+					return s.hbd, ctx.Err()
+				}
+			}
+		}
+
+		s.hbd.CurrentPage++
+		s.hbd.NextPageToken = resp.NextPageToken
+		s.hbd.ErrorCount += errorCount
+		s.hbd.ProcessedCount += len(resp.WorkflowExecutions)
```

Comment:
Would be great if these are domain specific metrics. To achieve that, you need to emit these metrics from within the task handler